core/src/main/java/org/apache/iceberg/hadoop/HadoopInputFile.java

```diff
@@ -133,7 +143,7 @@ public SeekableInputStream newStream() {
   }
 
   public Configuration getConf() {
-    return conf;
+    return fs.getConf();
```

Comment:
This needs to return the `Configuration` that was passed in when creating the `HadoopInputFile`, if there was one. The file system's `Configuration` may not contain the same configuration properties as the one that was passed in, and this `Configuration` is needed in some paths.

Reply:
@rdblue I made the change but would like to clarify. If `HadoopInputFile` `conf` may not be the same as `FileSystem` `conf`, will it be better to make it more explicit by implementing `Configurable` or extending from `Configured`?

Reply:
I don't think we want to add complexity to how `Configuration` objects are handled. A `Configuration` is used to instantiate a catalog, and it gets passed down to `FileIO` from there. We don't need to be able to change that configuration, we just need to make sure the one passed to create the catalog is used by parts like ORC that currently use a `Configuration`.

And if a `Configuration` is not passed, it is fine to use `new Configuration()` or the config from a file system that was passed. We don't make guarantees in that case.

Reply:
I don't see how extending from `Configured` will add complexity. The behavior of `HadoopInputFile` or `HadoopOutputFile` won't change.

Reply:
It's not a big change for Iceberg, but it would mean that there are multiple paths for passing a `Configuration` to parts that use it. I'd rather avoid that so that there is one way configurations are passed through code to avoid callers hijacking parts and updating the configuration held by a `HadoopOutputFile`.

Reply:
I don't see much difference between
```
Configuration conf = new Configuration();
Path path = new Path(...);
HadoopInputFile file = HadoopInputFile.fromPath(path, conf);
...
conf = new Configuration();
file.setConf(conf);
```
and
```
Configuration conf = new Configuration();
Path path = new Path(...);
HadoopInputFile file = HadoopInputFile.fromPath(path, conf);
...
conf.set();
```
if I correctly understand your concern with incorrectly updating the configuration. It would be necessary to copy passed configuration object to prevent hijacking.

Reply:
You can get input and output files by constructing them yourself, but the recommended way is to use the factory methods on a table's `FileIO` so that the catalog/table implementation can inject what it needs to at this layer, like using a different implementation for an object store.

The Hadoop variants of input and output file accept a configuration so that `HadoopFileIO` can pass it, but changing that `Configuration` is not a part of the API. We should avoid cases where implementations attempt to change the configuration that an `InputFile` or `OutputFile` uses. The right way to pass a configuration is to pass it in when creating a catalog.

Reply:
Still the same applies to `FileIO` implementation. Without a deep copy of `Configuration` inside `HadoopInputFile`, it is possible to cache `Configuration` object on a class that implements `FileIO` and expose methods that modify it. I do not suggest that it needs to be handled as part of this PR, just pointing out that even though it is not possible to change reference, the returned object is not immutable. Assuming that there are no other issues to address, can you please approve the PR?